drivers/soundwire/bus.c

```diff
@@ -1082,12 +1082,109 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
 	return ret;
 }
 
+static int sdw_slave_set_frequency(struct sdw_slave *slave)
+{
+	u32 mclk_freq = slave->bus->prop.mclk_freq;
+	u32 curr_freq = slave->bus->params.curr_dr_freq >> 1;
+	u8 base_offset = 0;
+	u8 base = 0;
+	u8 scale_index = 0;
+	u32 scale;
+	int ret;
+	int i;
+
+	/*
+	 * frequency base and scale registers are required for SDCA
+	 * devices. They may also be used for 1.2+/non-SDCA devices,
+	 * but we will need a DisCo property to cover this case
+	 */
+	if (!slave->id.class_id)
+		return 0;
+
+	if (!mclk_freq) {
+		dev_err(&slave->dev,
+			"no bus MCLK, cannot set SDW_SCP_BUS_CLOCK_BASE\n");
+		return -EINVAL;
+	}
+
+	/* map base frequency using Table 89 of SoundWire 1.2 spec */
+	if (!(38400000 % mclk_freq)) {
+		base = SDW_SCP_BASE_CLOCK_19200000_HZ;
+		base_offset = 1;
+	}
+	if (!(24000000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_24000000_HZ;
+	if (!(24576000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_24576000_HZ;
+	if (!(22579200 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_22579200_HZ;
+	if (!(32000000 % mclk_freq))
+		base = SDW_SCP_BASE_CLOCK_32000000_HZ;
+
+	/* paranoia checks */
+	if (!base) {
+		dev_err(&slave->dev,
+			"Unsupported clock base\n");
+		return -EINVAL;
+	}
+
+	if (mclk_freq % curr_freq) {
+		dev_err(&slave->dev,
+			"MCLK and bus curr_freq are not multiple\n");
+		return -EINVAL;
+	}
+
+	scale = mclk_freq / curr_freq - base_offset;
+
+	/* map scale to Table 90 of SoundWire 1.2 spec */
+	for (i = 0; i < 7; i++) {
+		if (scale == BIT(i)) {
+			scale_index = i + 1;
+			break;
+		}
+	}
```

Comment:
@plbossart can we use ffs() here?

Reply:
@ranj063 I thought about it but I really want to check for powers of 2 only. All other cases are errors. With ffs() you would get the first non-zero bit, not that there is a single bit set.

Reply:
I see. maybe is_power_of_2() then?

Reply:
Maybe. This is going to be multiple tests anyways since I also need to check the value is 64 or lower. The loop is more brute force but there's a single check.
I am sure @lyakh will have an optimization from hell on this one :-)

Reply:
I am eager for your reply. @lyakh

Reply:
Maybe
```
if (scale == BIT(bits_per(scale) - 1) && scale <= 64)
    scale_index = bits_per(scale);
else
    dev_err()
```
But it is less readable, so how about
```
switch (scale) {
    case 1:
    case 2:
    case 4:
    case 8:
    case 16:
    case 32:
    case 64:
        scale_index = bits_per(scale);
        break;
    default:
        dev_err()
        return -EINVAL;
```

Reply:
factor = ffs(scale);
if (! is_power_of_2 (scale) || factor >= 8)
            dev_err;

Reply:
@plbossart @RanderWang I feel flattered, but my otherwise direct super-low-latency link to the hell seems broken today :-) But no, I don't think I can much improve what already has been proposed. Something like this I guess:
```
scale_idx = ilog2(scale) + 1;
if (BIT(scale_idx - 1) != scale || scale_idx > 7) {
	/* error */
}
```
Using `ffs()` above instead of `ilog2() + 1` could be very slightly more optimal because usually the value would be within the low 7 bits but using `ilog2()` seems more explicit here, which is something we usually prefer ;-) Otherwise either of them should be fine.
Also, using one of these versions would eliminate the need to initialise `scale_idx`